monai/handlers/decollate_batch.py

```diff
@@ -0,0 +1,61 @@
+# Copyright 2020 - 2021 MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import TYPE_CHECKING
+
+from monai.config import IgniteInfo
+from monai.data import decollate_batch, rep_scalar_to_batch
+from monai.engines.utils import IterationEvents
+from monai.utils import min_version, optional_import
+
+Events, _ = optional_import("ignite.engine", IgniteInfo.OPT_IMPORT_VERSION, min_version, "Events")
+if TYPE_CHECKING:
+    from ignite.engine import Engine
+else:
+    Engine, _ = optional_import("ignite.engine", IgniteInfo.OPT_IMPORT_VERSION, min_version, "Engine")
+
+
+class DecollateBatch:
+    """
+    Ignite handler to execute the `decollate batch` logic for `engine.state.batch` and `engine.state.output`.
+    Typical usage is to set `decollate=False` in the engine and execute some postprocessing logic first
+    then decollate the batch, otherwise, engine will decollate batch before the postprocessing.
+
+    Args:
+        event: expected EVENT to attach the handler, should be "MODEL_COMPLETED" or "ITERATION_COMPLETED".
+            default to "MODEL_COMPLETED".
+
+    """
+
+    def __init__(self, event: str = "MODEL_COMPLETED"):
+        event = event.upper()
+        if event not in ("MODEL_COMPLETED", "ITERATION_COMPLETED"):
+            raise ValueError("event should be `MODEL_COMPLETED` or `ITERATION_COMPLETED`.")
+        self.event = event
+
+    def attach(self, engine: Engine) -> None:
+        """
+        Args:
+            engine: Ignite Engine, it can be a trainer, validator or evaluator.
+        """
+        if self.event == "MODEL_COMPLETED":
+            engine.add_event_handler(IterationEvents.MODEL_COMPLETED, self)
+        else:
+            engine.add_event_handler(Events.ITERATION_COMPLETED, self)
+
+    def __call__(self, engine: Engine) -> None:
+        """
+        Args:
+            engine: Ignite Engine, it can be a trainer, validator or evaluator.
+        """
+        # replicate the scalar values to make sure all the items have batch dimension, then decollate
+        engine.state.batch = decollate_batch(rep_scalar_to_batch(engine.state.batch), detach=True)
```

Comment:
I think `rep_scalar_to_batch` and `detach=True` should be configurable,
whether to apply to `batch` or `output` or both should also be configurable.

not sure if we want to support the decollate of a specific dictionary key if `batch` and `output` are dictionaries

Reply:
Thanks for your suggestions.
Actually, when I was developing `rep_scalar_to_batch`, I also thought about only decollating several keys of the dict data. But I didn't get how to save the results, currently, we decollate a dict to a list of dicts. Any more thoughts about how to save the results?

Thanks.

Reply:
OK, I can add this logic to specify keys. Seems this is the best way so far, I will try to make it clear in the doc-string that it will **discard the data** with keys not in the list. It's different from other dict transforms which always keep the not specified keys.
Anyway, as we just released it in v0.6, we can keep refactoring the `decollate` design according to users' feedback, like what we did for the `Invertd` feature.

Thanks.

Reply:
Hi @wyli ,

I feel maybe we don't need `rep_scalar_to_batch` to be configurable, because it will fail if decollating data with scalar tensors and didn't specify the keys.
What do you think?

Thanks.

Reply:
updated in the latest commit.
Thanks.

Reply:
sure @Nic-Ma, we can keep it hard-coded, maybe in the future we will have cases where we don't want to replicate the scalars. I agree that we can refactor it later...

Reply:
Thanks, removed `rep_scalar` arg in the latest commit.